File: RobloxAddon/src/XBedrock.lua
local XBedrock = {}
XBedrock.__index = XBedrock

-- Services
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Players = game:GetService("Players")
local RunService = game:GetService("RunService")
local HttpService = game:GetService("HttpService")

-- Constants
local DEFAULT_CONFIG = {
    ServerIP = "localhost",
    ServerPort = 19132,
    SyncInterval = 1,
    EnableChat = true,
    EnableInventory = true,
    EnableCosmetics = true
}

-- Private variables
local _config = {}
local _connected = false
local _lastSync = 0
local _webSocket = nil
local _eventHandlers = {}

-- Event handlers
local Events = {
    Chat = Instance.new("BindableEvent"),
    Position = Instance.new("BindableEvent"),
    Inventory = Instance.new("BindableEvent"),
    Cosmetics = Instance.new("BindableEvent")
}

-- Utility functions
local function createWebSocket()
    local success, result = pcall(function()
        return WebSocket.connect("ws://" .. _config.ServerIP .. ":" .. _config.ServerPort)
    end)
    
    if success then
        return result
    else
        warn("Failed to create WebSocket connection:", result)
        return nil
    end
end

local function handleMessage(message)
    local data = HttpService:JSONDecode(message)
    
    if data.type == "chat" and _config.EnableChat then
        Events.Chat:Fire(data.message)
    elseif data.type == "position" then
        Events.Position:Fire(data.position)
    elseif data.type == "inventory" and _config.EnableInventory then
        Events.Inventory:Fire(data.items)
    elseif data.type == "cosmetics" and _config.EnableCosmetics then
        Events.Cosmetics:Fire(data.cosmetics)
    end
end

-- Public methods
function XBedrock.new(config)
    local self = setmetatable({}, XBedrock)
    _config = table.clone(DEFAULT_CONFIG)
    
    -- Merge custom config
    for key, value in pairs(config or {}) do
        _config[key] = value
    end
    
    return self
end

function XBedrock:Initialize(config)
    if config then
        for key, value in pairs(config) do
            _config[key] = value
        end
    end
    
    -- Set up WebSocket connection
    _webSocket = createWebSocket()
    if _webSocket then
        _connected = true
        
        -- Set up message handler
        _webSocket.OnMessage:Connect(handleMessage)
        
        -- Set up connection closed handler
        _webSocket.OnClose:Connect(function()
            _connected = false
            -- Attempt to reconnect after 5 seconds
            task.delay(5, function()
                self:Initialize()
            end)
        end)
    end
    
    -- Set up position sync
    if _config.SyncInterval > 0 then
        RunService.Heartbeat:Connect(function()
            local now = tick()
            if now - _lastSync >= _config.SyncInterval then
                _lastSync = now
                self:SendPosition(Players.LocalPlayer.Character and 
                    Players.LocalPlayer.Character:GetPivot().Position or Vector3.new(0, 0, 0))
            end
        end)
    end
end

function XBedrock:Connect()
    if not _connected then
        self:Initialize()
    end
end

function XBedrock:Disconnect()
    if _webSocket then
        _webSocket:Close()
        _webSocket = nil
        _connected = false
    end
end

function XBedrock:SendPosition(position)
    if _connected and _webSocket then
        local data = {
            type = "position",
            position = {
                x = position.X,
                y = position.Y,
                z = position.Z
            }
        }
        _webSocket:Send(HttpService:JSONEncode(data))
    end
end

function XBedrock:SendChat(message)
    if _connected and _webSocket and _config.EnableChat then
        local data = {
            type = "chat",
            message = message
        }
        _webSocket:Send(HttpService:JSONEncode(data))
    end
end

function XBedrock:SendInventory(items)
    if _connected and _webSocket and _config.EnableInventory then
        local data = {
            type = "inventory",
            items = items
        }
        _webSocket:Send(HttpService:JSONEncode(data))
    end
end

function XBedrock:SendCosmetics(cosmetics)
    if _connected and _webSocket and _config.EnableCosmetics then
        local data = {
            type = "cosmetics",
            cosmetics = cosmetics
        }
        _webSocket:Send(HttpService:JSONEncode(data))
    end
end

-- Event handlers
function XBedrock:OnChat(callback)
    Events.Chat.Event:Connect(callback)
end

function XBedrock:OnPosition(callback)
    Events.Position.Event:Connect(callback)
end

function XBedrock:OnInventory(callback)
    Events.Inventory.Event:Connect(callback)
end

function XBedrock:OnCosmetics(callback)
    Events.Cosmetics.Event:Connect(callback)
end

return XBedrock 

File: browser/xbedrock.js
class XBedrock {
    constructor(config = {}) {
        this.config = {
            serverUrl: config.serverUrl || 'ws://localhost:8080',
            syncInterval: config.syncInterval || 1000,
            features: {
                chat: config.features?.chat ?? true,
                inventory: config.features?.inventory ?? true,
                cosmetics: config.features?.cosmetics ?? true,
                webstore: config.features?.webstore ?? true
            }
        };

        this.connected = false;
        this.ws = null;
        this.playerData = {
            position: { x: 0, y: 0, z: 0 },
            inventory: [],
            cosmetics: [],
            webstore: {}
        };

        this.eventHandlers = {
            chat: [],
            position: [],
            inventory: [],
            cosmetics: [],
            webstore: [],
            connect: [],
            disconnect: [],
            error: []
        };
    }

    // Connect to server
    connect() {
        if (this.connected) return;

        this.ws = new WebSocket(this.config.serverUrl);

        this.ws.onopen = () => {
            this.connected = true;
            this._triggerEvent('connect');
            this._startSyncLoop();
        };

        this.ws.onclose = () => {
            this.connected = false;
            this._triggerEvent('disconnect');
        };

        this.ws.onerror = (error) => {
            this._triggerEvent('error', error);
        };

        this.ws.onmessage = (event) => {
            try {
                const data = JSON.parse(event.data);
                this._handleMessage(data);
            } catch (error) {
                this._triggerEvent('error', error);
            }
        };
    }

    // Disconnect from server
    disconnect() {
        if (!this.connected) return;

        this.ws.close();
        this.connected = false;
    }

    // Send chat message
    sendChat(message) {
        if (!this.config.features.chat) return;
        this._send('chat', { message });
    }

    // Send position
    sendPosition(position) {
        this._send('position', { position });
    }

    // Update inventory
    updateInventory(inventory) {
        if (!this.config.features.inventory) return;
        this._send('inventory', { inventory });
    }

    // Update cosmetics
    updateCosmetics(cosmetics) {
        if (!this.config.features.cosmetics) return;
        this._send('cosmetics', { cosmetics });
    }

    // Update webstore
    updateWebstore(webstore) {
        if (!this.config.features.webstore) return;
        this._send('webstore', { webstore });
    }

    // Get player data
    getPlayerData() {
        return { ...this.playerData };
    }

    // Event handling
    on(event, handler) {
        if (this.eventHandlers[event]) {
            this.eventHandlers[event].push(handler);
        }
    }

    off(event, handler) {
        if (this.eventHandlers[event]) {
            this.eventHandlers[event] = this.eventHandlers[event].filter(h => h !== handler);
        }
    }

    // Private methods
    _send(type, data) {
        if (!this.connected) return;

        this.ws.send(JSON.stringify({
            type,
            data
        }));
    }

    _handleMessage(message) {
        const { type, data } = message;

        switch (type) {
            case 'chat':
                if (this.config.features.chat) {
                    this._triggerEvent('chat', data);
                }
                break;

            case 'position':
                this.playerData.position = data.position;
                this._triggerEvent('position', data);
                break;

            case 'inventory':
                if (this.config.features.inventory) {
                    this.playerData.inventory = data.inventory;
                    this._triggerEvent('inventory', data);
                }
                break;

            case 'cosmetics':
                if (this.config.features.cosmetics) {
                    this.playerData.cosmetics = data.cosmetics;
                    this._triggerEvent('cosmetics', data);
                }
                break;

            case 'webstore':
                if (this.config.features.webstore) {
                    this.playerData.webstore = data.webstore;
                    this._triggerEvent('webstore', data);
                }
                break;
        }
    }

    _triggerEvent(event, data) {
        if (this.eventHandlers[event]) {
            this.eventHandlers[event].forEach(handler => handler(data));
        }
    }

    _startSyncLoop() {
        setInterval(() => {
            if (!this.connected) return;

            // Sync position
            this.sendPosition(this.playerData.position);

            // Sync inventory
            if (this.config.features.inventory) {
                this.updateInventory(this.playerData.inventory);
            }

            // Sync cosmetics
            if (this.config.features.cosmetics) {
                this.updateCosmetics(this.playerData.cosmetics);
            }

            // Sync webstore
            if (this.config.features.webstore) {
                this.updateWebstore(this.playerData.webstore);
            }
        }, this.config.syncInterval);
    }
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = XBedrock;
} else if (typeof define === 'function' && define.amd) {
    define([], function() { return XBedrock; });
} else {
    window.XBedrock = XBedrock;
} 